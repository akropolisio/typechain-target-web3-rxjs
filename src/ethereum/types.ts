import { EventLog as Web3EventLog, TransactionReceipt } from 'web3-core';
import BN from 'bn.js';

export type EventLog<T> = Omit<Web3EventLog, 'returnValues'> & { returnValues: T };

export interface EventEmitter<T> {
  on(type: 'data', handler: (event: EventLog<T>) => void): EventEmitter<T>;
  on(type: 'changed', handler: (receipt: EventLog<T>) => void): EventEmitter<T>;
  on(type: 'error', handler: (error: Error) => void): EventEmitter<T>;
  on(
    type: 'error' | 'data' | 'changed',
    handler: (error: Error | TransactionReceipt | string) => void,
  ): EventEmitter<T>;
}

export interface Tx {
  nonce?: string | number;
  chainId?: string | number;
  from?: string;
  to?: string;
  data?: string;
  value?: string | number;
  gas?: string | number;
  gasPrice?: string | number;
  maxPriorityFeePerGas?: string | number;
  maxFeePerGas?: string | number;
}

export type InputEvmTypeToJSTypeMap = {
  address: string;
  integer: BN;
  uinteger: BN;
  boolean: boolean;
  string: string;
  bytes: string;
  'dynamic-bytes': string;
};

export type InputEvmType = keyof InputEvmTypeToJSTypeMap;

export type JSType = InputEvmTypeToJSTypeMap[InputEvmType];
