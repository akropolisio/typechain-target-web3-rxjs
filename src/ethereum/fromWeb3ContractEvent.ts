import { fromEventPattern, Observable } from 'rxjs';
import { EventData } from 'web3-eth-contract';

type EventEmitter = {
  on(event: string, handler: (event: EventData) => void): EventEmitter;
};

export function fromWeb3DataEvent(emitter: EventEmitter): Observable<EventData> {
  interface IUnsubscribable {
    unsubscribe: () => void;
  }

  return fromEventPattern<EventData>(
    handler => emitter.on('data', handler),
    (_, signal: IUnsubscribable) => signal.unsubscribe(),
  );
}
