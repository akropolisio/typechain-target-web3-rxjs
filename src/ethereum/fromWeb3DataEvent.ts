import { fromEventPattern, Observable } from 'rxjs';
import { EventLog } from 'web3-core';

import { EventEmitter } from './types';

export function fromWeb3DataEvent(emitter: EventEmitter<any>): Observable<EventLog> {
  interface IUnsubscribable {
    unsubscribe: () => void;
  }

  return fromEventPattern<EventLog>(
    handler => emitter.on('data', handler as (event: EventLog) => void),
    (_, signal: IUnsubscribable) => signal.unsubscribe(),
  );
}
