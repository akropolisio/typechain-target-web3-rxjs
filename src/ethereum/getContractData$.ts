import { Observable, from, merge, EMPTY, ReplaySubject } from 'rxjs';
import { skipUntil, mergeMap, throttleTime, switchMap, shareReplay } from 'rxjs/operators';
import { Eth } from 'web3-eth';
import { Contract } from 'web3-eth-contract';

import {
  EventLog,
  CallOptions,
  Web3ContractInputArgs,
  Web3ContractResponse,
  Response,
} from './types';

interface IOptions {
  eventsForReload?: Observable<EventLog<any>> | Observable<EventLog<any>>[];
  reloadTrigger$?: Observable<any>;
  args?: Web3ContractInputArgs;
  convert?(value: Web3ContractResponse): Response | void;
  tx?: CallOptions;
}

function identity(value: any) {
  return value;
}

function awaitMs(delayMs: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, delayMs));
}

export function getContractData$(
  contract: Contract,
  eth: Eth,
  method: string,
  options: IOptions = {},
): Observable<Response> {
  const { eventsForReload = [], reloadTrigger$ = EMPTY, args = [], convert = identity } = options;

  const load = async () => {
    const data = await contract.methods[method](...args).call(options.tx);
    return convert(data);
  };

  const reloadEvents = Array.isArray(eventsForReload) ? eventsForReload : [eventsForReload];

  const first$ = from(load());
  const fromEvents$ = merge(...reloadEvents).pipe(
    skipUntil(first$),
    throttleTime(200),
    switchMap(async event => {
      let currentBlock = await eth.getBlockNumber();

      /* eslint-disable no-await-in-loop */
      while (currentBlock < event.blockNumber) {
        await awaitMs(500);
        currentBlock = await eth.getBlockNumber();
      }
      /* eslint-enable no-await-in-loop */

      return event;
    }),
    mergeMap(() => from(load()), 1),
    shareReplay(1),
  );

  const subject = new ReplaySubject<Response>(1);

  merge(first$, fromEvents$, reloadTrigger$).subscribe(subject);

  return subject;
}
