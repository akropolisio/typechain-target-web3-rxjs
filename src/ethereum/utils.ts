import { RawAbiDefinition, RawAbiParameter } from 'typechain';

export function extractAbiMethod(abi: RawAbiDefinition[], method: string): RawAbiDefinition | null {
  const [, methodName] = method.match(/(\w+?)($|\(.*)/) || [];

  if (!methodName) {
    console.warn(new Error(`Invalid method: ${method}`));
    return null;
  }

  const methodEntries = abi.filter(m => m.name === methodName && m.type === 'function');

  if (!methodEntries.length) {
    console.warn(new Error(`Method ${method} not found`));
    return null;
  }

  const isSimpleMethodName = method === methodName;
  if (isSimpleMethodName) {
    return methodEntries[0];
  }

  const entry = methodEntries.find(m => getWeb3Signature(m) === method);

  if (!entry) {
    console.warn(new Error(`Method ${method} not found`));
    return null;
  }

  return entry;
}

function getWeb3Signature(abi: Pick<RawAbiDefinition, 'name' | 'inputs'>): string {
  return `${abi.name}(${abi.inputs.map(i => getArgumentForSignature(i)).join(',')})`;
}

function getArgumentForSignature(argument: RawAbiParameter): string {
  if (argument.type === 'tuple') {
    return `(${argument.components?.map(i => getArgumentForSignature(i)).join(',')})`;
  }
  if (argument.type === 'tuple[]') {
    return `${getArgumentForSignature({ ...argument, type: 'tuple' })}[]`;
  }
  return argument.type;
}

export function attachStaticFields<T extends {}, I extends Record<string, any>>(
  target: T,
  staticFields: I,
): T & I {
  const result: T & I = target as T & I;

  Object.keys(staticFields).forEach((key: keyof I) => {
    (result as I)[key] = staticFields[key];
  });

  return result;
}
