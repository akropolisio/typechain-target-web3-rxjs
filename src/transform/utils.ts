import { O } from 'ts-toolbelt';
import {
  AbiOutputParameter,
  AbiParameter,
  EventDeclaration,
  EvmOutputType,
  FunctionDeclaration,
  getSignatureForFn,
  TupleType,
} from 'typechain';

export function unfoldDeclarations<T extends FunctionDeclaration | EventDeclaration>(
  declaration: Record<string, T[]>,
): T[] {
  const toOverloadedDeclaration = (entry: T) => ({
    ...entry,
    name: `"${getSignatureForFn(entry as FunctionDeclaration)}"`,
  });

  return Object.values(declaration).flatMap(entries =>
    entries.length === 1 ? entries[0] : [entries[0], ...entries.map(toOverloadedDeclaration)],
  );
}

export function codegenInputTypes(inputs: O.Optional<AbiParameter, 'name'>[]): string {
  if (inputs.length === 0) {
    return 'void';
  }
  return `{
    ${inputs
      .map((input, index) => `${input.name || `arg${index}`}: ${codegenType(input.type)}`)
      .join(';\n')}
  }`;
}

export function codegenOutputTypes(outputs: O.Optional<AbiOutputParameter, 'name'>[]): string {
  if (outputs.length === 1) {
    return codegenType(outputs[0].type);
  }
  return `{
      ${outputs.map(t => t.name && `${t.name}: ${codegenType(t.type)};\n`).join('')}
      ${outputs.map((t, i) => `${i}: ${codegenType(t.type)}`).join(';\n')}
    }`;
}

export function codegenType(evmType: EvmOutputType): string {
  switch (evmType.type) {
    case 'integer':
    case 'uinteger':
      return 'BN';
    case 'address':
      return 'string';
    case 'bytes':
    case 'dynamic-bytes':
      return 'string';
    case 'array':
      return `(${codegenType(evmType.itemType)})[]`;
    case 'boolean':
      return 'boolean';
    case 'string':
      return 'string';
    case 'tuple':
      return codegenTupleType(evmType);
    case 'void':
      return 'void';
    case 'unknown':
    default:
      throw new Error(`Unhandled type while code generation: ${evmType.originalType}`);
  }
}

export function codegenTupleType(tuple: TupleType) {
  return `{
    ${tuple.components
      .map(component => `${component.name}: ${codegenType(component.type)}`)
      .join(';\n')}
  }`;
}

export function uniq<T>(array: T[]): T[] {
  return array.reduce((acc, curr) => (acc.includes(curr) ? acc : [...acc, curr]), [] as T[]);
}
