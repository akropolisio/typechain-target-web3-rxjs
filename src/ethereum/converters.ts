import BN from 'bn.js';
import { RawAbiParameter } from 'typechain';

import {
  JSType,
  EvmType,
  Arguments,
  Response,
  InputEvmTypeToJSTypeMap,
  OutputEvmTypeToJSTypeMap,
  Web3ContractInputArgs,
  Web3ContractResponse,
} from './types';
import { attachStaticFields } from './utils';

export function inputsToRequest(
  inputsAbi: RawAbiParameter[],
  args: Arguments,
): Web3ContractInputArgs {
  if (inputsAbi.length === 0 && args) {
    throw new Error(`Received arguments ${args} for empty ABI declaration`);
  }

  return inputsAbi.map(parameterAbi => {
    const { isArray, type } = parseEvmType(parameterAbi.type);
    const arg = args[parameterAbi.name];
    if (!arg) {
      throw new Error(`ABI parameter ${parameterAbi.name} not found in ${args}`);
    }
    if (type === 'tuple') {
      return isArray
        ? (arg as Arguments[]).map(innerArgs =>
            inputsToRequest(parameterAbi.components || [], innerArgs),
          )
        : inputsToRequest(parameterAbi.components || [], arg as Arguments);
    }

    return isArray
      ? (arg as JSType[]).map(innerArgs => convertInputValueToRequest(type, innerArgs))
      : convertInputValueToRequest(type, arg as JSType);
  });
}

function convertInputValueToRequest(type: EvmType, inputValue: JSType): JSType {
  const toRequest: {
    [key in EvmType]: (input: InputEvmTypeToJSTypeMap[key]) => JSType;
  } = {
    address: value => value,
    boolean: value => value,
    integer: value => value.toString(),
    uinteger: value => value.toString(),
    string: value => value,
    bytes: value => value,
    'dynamic-bytes': value => value,
  };

  return (toRequest[type] as any)(inputValue);
}

export function responseToOutput(
  outputsAbi: RawAbiParameter[],
  response: Web3ContractResponse,
): Response {
  if (outputsAbi.length === 0 && response) {
    throw new Error(`Received response ${response} for empty ABI declaration`);
  }
  const responseArray = Array.isArray(response) ? response : [response];

  const result = outputsAbi.reduce((acc, parameterAbi, index) => {
    const { isArray, type } = parseEvmType(parameterAbi.type);
    const responseItem = responseArray[index];

    if (!responseItem) {
      throw new Error(`ABI parameter ${parameterAbi.name} not found in ${response}`);
    }

    let output;
    if (type === 'tuple') {
      output = isArray
        ? (responseItem as JSType[]).map(innerValues =>
            responseToOutput(parameterAbi.components || [], innerValues),
          )
        : responseToOutput(parameterAbi.components || [], responseItem);
    } else {
      output = isArray
        ? (responseItem as JSType[]).map(innerValues =>
            convertResponseValueToOutput(type, innerValues),
          )
        : convertResponseValueToOutput(type, responseItem as JSType);
    }

    return attachStaticFields([...acc, output], {
      [parameterAbi.name]: output,
    });
  }, [] as Response[]);

  return result.length === 1 ? result[0] : result;
}

function convertResponseValueToOutput(type: EvmType, responseValue: JSType): JSType {
  const toOutput: {
    [key in EvmType]: (response: OutputEvmTypeToJSTypeMap[key]) => JSType;
  } = {
    address: value => value,
    boolean: value => value,
    integer: value => new BN(value),
    uinteger: value => new BN(value),
    string: value => value,
    bytes: value => value,
    'dynamic-bytes': value => value,
  };

  return (toOutput[type] as any)(responseValue);
}

function parseEvmType(abiType: string): { type: EvmType | 'tuple'; isArray: boolean } {
  const [, inputType, isArray] = abiType.match(/^(.+?)(\[\])?$/) || [];

  const type = ((): EvmType | 'tuple' => {
    // eslint-disable-next-line default-case
    switch (abiType) {
      case 'bool':
        return 'boolean';
      case 'address':
        return 'address';
      case 'string':
        return 'string';
      case 'byte':
        return 'bytes';
      case 'bytes':
        return 'dynamic-bytes';
    }

    if (inputType.startsWith('tuple')) {
      return 'tuple';
    }
    if (inputType.startsWith('uint')) {
      return 'uinteger';
    }
    if (inputType.startsWith('int')) {
      return 'integer';
    }
    if (inputType.startsWith('bytes')) {
      return 'bytes';
    }

    throw new Error(`Type "${abiType}" cannot be parsed`);
  })();

  return { type, isArray: !!isArray };
}
