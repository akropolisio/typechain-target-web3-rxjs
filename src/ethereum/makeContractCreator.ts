import Web3 from 'web3';
import { Contract } from 'web3-eth-contract';

import { getContractData$ } from './getContractData$';
import {
  EventEmitter,
  SendOptions,
  CallOptions,
  ContractWrapper,
  CallMethod,
  ReadMethod,
  SendMethod,
  Arguments,
} from './types';
import { attachStaticFields, extractAbiMethod } from './utils';
import { inputsToRequest, responseToOutput } from './converters';

export function makeContractCreator<C extends ContractWrapper>(abi: any[]) {
  return (web3: Web3, address: string): C => {
    const baseContract = new web3.eth.Contract(abi, address);

    const methodsProxy = new Proxy<Contract['methods']>(
      {},
      {
        get(target: Contract['methods'], prop: string) {
          const methodAbi = extractAbiMethod(abi, prop);

          if (!methodAbi) {
            return target[prop];
          }

          const baseCallFunction = (
            input: void | Arguments,
            eventsForReload?: EventEmitter<any> | EventEmitter<any>[],
            tx?: CallOptions,
          ) => {
            return getContractData$(baseContract, web3.eth, prop, {
              tx,
              args: input ? inputsToRequest(methodAbi.inputs, input) : [],
              eventsForReload,
              convert: result => (result ? responseToOutput(methodAbi.outputs, result) : undefined),
            });
          };

          const callFunction: CallMethod<any, any> = baseCallFunction;

          const readFunction: ReadMethod<any, any> = (
            input: void | Arguments,
            tx: SendOptions,
            eventsForReload?: EventEmitter<any> | EventEmitter<any>[],
          ) => baseCallFunction(input, eventsForReload, tx);

          if (methodAbi.stateMutability === 'view') {
            return callFunction;
          }

          const getTransactionFunction = (input: void | Arguments) => {
            const args = input ? inputsToRequest(methodAbi.inputs, input) : [];
            return baseContract.methods[prop](...args);
          };

          const sendFunction: SendMethod<any, any> = attachStaticFields(
            (input: void | Arguments, tx: SendOptions) => getTransactionFunction(input).send(tx),
            {
              getTransaction: getTransactionFunction,
              read: readFunction,
            },
          );

          return sendFunction;
        },
      },
    );

    return (new Proxy<Contract>(baseContract, {
      get(target, prop: keyof Contract) {
        if (prop === 'methods') {
          return methodsProxy;
        }
        return target[prop];
      },
    }) as unknown) as C;
  };
}
