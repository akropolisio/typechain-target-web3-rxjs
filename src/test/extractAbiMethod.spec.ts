import { extractAbiMethod } from 'ethereum/utils';

import testAbi from './abi/testAbi.json';

beforeAll(() => {
  jest.spyOn(console, 'warn').mockImplementation(() => {});
});

describe('extractAbiMethod', () => {
  it('should return ABI for simple method', () => {
    const method = 'testSend';
    expect({ method, result: extractAbiMethod(testAbi as any, method) }).toMatchSnapshot();
  });

  it('should return ABI for overloaded method', () => {
    const method = 'testRead((bool,address[2],(uint256,uint256)[]),(bytes,string),uint256)';
    expect({ method, result: extractAbiMethod(testAbi as any, method) }).toMatchSnapshot();
  });

  it('should return first ABI for overloaded methods if simple name used', () => {
    const method = 'testRead';
    expect({ method, result: extractAbiMethod(testAbi as any, method) }).toMatchSnapshot();
  });

  it('should return null when method not found', () => {
    const method = 'unknownMethod';
    expect({ method, result: extractAbiMethod(testAbi as any, method) }).toMatchSnapshot();
  });
});
