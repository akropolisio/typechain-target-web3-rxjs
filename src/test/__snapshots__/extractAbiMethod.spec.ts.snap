// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractAbiMethod should return ABI for overloaded method 1`] = `
Object {
  "method": "testRead((bool,address[],(uint256,uint256)[]),(bytes,string),uint256)",
  "result": Object {
    "inputs": Array [
      Object {
        "components": Array [
          Object {
            "name": "a",
            "type": "bool",
          },
          Object {
            "name": "b",
            "type": "address[]",
          },
          Object {
            "components": Array [
              Object {
                "name": "x",
                "type": "uint256",
              },
              Object {
                "name": "y",
                "type": "uint256",
              },
            ],
            "name": "c",
            "type": "tuple[]",
          },
        ],
        "name": "s",
        "type": "tuple",
      },
      Object {
        "components": Array [
          Object {
            "name": "x",
            "type": "bytes",
          },
          Object {
            "name": "y",
            "type": "string",
          },
        ],
        "name": "t",
        "type": "tuple",
      },
      Object {
        "name": "a",
        "type": "uint256",
      },
    ],
    "name": "testRead",
    "outputs": Array [
      Object {
        "name": "",
        "type": "address",
      },
      Object {
        "name": "Test",
        "type": "address",
      },
      Object {
        "name": "",
        "type": "bool",
      },
    ],
    "stateMutability": "view",
    "type": "function",
  },
}
`;

exports[`extractAbiMethod should return ABI for simple method 1`] = `
Object {
  "method": "testSend",
  "result": Object {
    "inputs": Array [
      Object {
        "components": Array [
          Object {
            "name": "a",
            "type": "uint256",
          },
          Object {
            "name": "b",
            "type": "uint256[]",
          },
          Object {
            "components": Array [
              Object {
                "name": "x",
                "type": "uint256",
              },
              Object {
                "name": "y",
                "type": "uint256",
              },
            ],
            "name": "c",
            "type": "tuple[]",
          },
        ],
        "name": "s",
        "type": "tuple",
      },
      Object {
        "components": Array [
          Object {
            "name": "x",
            "type": "uint256",
          },
          Object {
            "name": "y",
            "type": "uint256",
          },
        ],
        "name": "t",
        "type": "tuple",
      },
      Object {
        "name": "a",
        "type": "uint256",
      },
    ],
    "name": "testSend",
    "outputs": Array [
      Object {
        "name": "",
        "type": "bool",
      },
    ],
    "stateMutability": "nonpayable",
    "type": "function",
  },
}
`;

exports[`extractAbiMethod should return first ABI for overloaded methods if simple name used 1`] = `
Object {
  "method": "testRead",
  "result": Object {
    "inputs": Array [],
    "name": "testRead",
    "outputs": Array [
      Object {
        "name": "",
        "type": "uint256",
      },
    ],
    "stateMutability": "view",
    "type": "function",
  },
}
`;

exports[`extractAbiMethod should return null when method not found 1`] = `
Object {
  "method": "unknownMethod",
  "result": null,
}
`;
